lib/spack/llnl/util/symlink.py

```diff
@@ -2,77 +2,167 @@
+            raise SymlinkError("An exception occurred while creating symlink") from e
+
+    # Redundancy check to make sure the link created successfully
+    if os.path.lexists(link_path):
+        if not os.path.exists(link_path):
+            # This is a broken link.
+            raise SymlinkError(
+                f"Broken link does not point to the source path ({source_path}).",
+                long_message="This can be caused by the source path not being relative to "
+                "either the link's parent directory or the current working directory or "
+                "the source target not existing.",
+            )
+    else:
+        raise SymlinkError("Link does not exist after symlink methods finished.")
+
+
+def islink(path: str) -> bool:
+    """Override os.islink to give correct answer for spack logic.
+
+    For Non-Windows: a link can be determined with the os.path.islink method.
+    Windows-only methods will return false for other operating systems.
 
+    For Windows: spack considers symlinks, hard links, and junctions to
+    all be links, so if any of those are True, return True.
 
-def islink(path):
-    return os.path.islink(path) or _win32_is_junction(path)
+    Args:
+        path (str): path to check if it is a link.
 
+    Returns:
+         bool - whether the path is any kind link or not.
+    """
+    try:
+        return any([os.path.islink(path), _windows_is_junction(path), _windows_is_hardlink(path)])
+    except Exception as e:
+        raise SymlinkError("Could not determine if given path is a link") from e
 
-# '_win32' functions based on
-# https://github.com/Erotemic/ubelt/blob/master/ubelt/util_links.py
-def _win32_junction(path, link):
-    # junctions require absolute paths
-    if not os.path.isabs(link):
-        link = os.path.abspath(link)
 
-    # os.symlink will fail if link exists, emulate the behavior here
-    if exists(link):
-        raise OSError(errno.EEXIST, "File  exists: %s -> %s" % (link, path))
+def _windows_is_hardlink(path: str) -> bool:
+    """Determines if a path is a windows hard link. This is accomplished
+    by looking at the number of links using os.stat. A non-hard-linked file
+    will have a st_nlink value of 1, whereas a hard link will have a value
+    larger than 1. Note that both the original and hard-linked file will
+    return True because they share the same inode.
 
-    if not os.path.isabs(path):
-        parent = os.path.join(link, os.pardir)
-        path = os.path.join(parent, path)
-        path = os.path.abspath(path)
+    Args:
+        path (str): Windows path to check for a hard link
+
+    Returns:
+         bool - Whether the path is a hard link or not.
+    """
+    if sys.platform != "win32" or os.path.islink(path) or not os.path.exists(path):
+        return False
+
+    try:
+        return os.stat(path).st_nlink > 1
+    except Exception as e:
+        raise SymlinkError("Could not determine if path is a hard link") from e
```

Comment:
I suppose I thought it would be easier to debug if the original exception was annotated, though that message does not particularly add any insight to the failure. I will take it out.